File: behavioral-design-pattern/state-pattern/state.py
"""
State Pattern
=================
Category: Behavioral Design Pattern

Intent:
    Allow an object to alter its behavior when its internal state changes.
    The object will appear to change its class.

When to use:
    - When an object's behavior depends on its state and it must change
      behavior at runtime depending on that state.
    - When you want to eliminate large conditional statements that select
      behavior based on the current state.
    - When transitions between states need to be explicit and controlled.

Key Participants:
    - State (interface): DocumentState -- declares methods that all concrete
      states must implement.
    - ConcreteState: DraftState, ReviewState, PublishedState, ArchivedState --
      each implements behavior appropriate for that state of the context.
    - Context: Document -- maintains a reference to a ConcreteState subclass
      that defines the current state.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


# ---------------------------------------------------------------------------
# State interface
# ---------------------------------------------------------------------------
class DocumentState(ABC):
    """Abstract base class for all document states."""

    @abstractmethod
    def name(self) -> str: ...

    @abstractmethod
    def edit(self, doc: Document) -> None: ...

    @abstractmethod
    def submit(self, doc: Document) -> None: ...

    @abstractmethod
    def publish(self, doc: Document) -> None: ...

    @abstractmethod
    def archive(self, doc: Document) -> None: ...


# ---------------------------------------------------------------------------
# Concrete states
# ---------------------------------------------------------------------------
class DraftState(DocumentState):
    """The document is being authored and can be freely edited."""

    def name(self) -> str:
        return "Draft"

    def edit(self, doc: Document) -> None:
        print("  [Draft] Editing document content.")

    def submit(self, doc: Document) -> None:
        print("  [Draft] Submitting document for review.")
        doc.set_state(ReviewState())

    def publish(self, doc: Document) -> None:
        print("  [Draft] ERROR: Cannot publish directly from Draft. Submit for review first.")

    def archive(self, doc: Document) -> None:
        print("  [Draft] Archiving draft document.")
        doc.set_state(ArchivedState())


class ReviewState(DocumentState):
    """The document is under editorial review."""

    def name(self) -> str:
        return "Under Review"

    def edit(self, doc: Document) -> None:
        print("  [Review] Sending document back to Draft for edits.")
        doc.set_state(DraftState())

    def submit(self, doc: Document) -> None:
        print("  [Review] ERROR: Document is already under review.")

    def publish(self, doc: Document) -> None:
        print("  [Review] Review approved. Publishing document.")
        doc.set_state(PublishedState())

    def archive(self, doc: Document) -> None:
        print("  [Review] ERROR: Cannot archive while under review. Publish or reject first.")


class PublishedState(DocumentState):
    """The document is live and visible to readers."""

    def name(self) -> str:
        return "Published"

    def edit(self, doc: Document) -> None:
        print("  [Published] ERROR: Cannot edit a published document. Archive it first.")

    def submit(self, doc: Document) -> None:
        print("  [Published] ERROR: Document is already published.")

    def publish(self, doc: Document) -> None:
        print("  [Published] ERROR: Document is already published.")

    def archive(self, doc: Document) -> None:
        print("  [Published] Archiving published document.")
        doc.set_state(ArchivedState())


class ArchivedState(DocumentState):
    """The document has been archived and is read-only."""

    def name(self) -> str:
        return "Archived"

    def edit(self, doc: Document) -> None:
        print("  [Archived] Un-archiving and moving to Draft for editing.")
        doc.set_state(DraftState())

    def submit(self, doc: Document) -> None:
        print("  [Archived] ERROR: Cannot submit an archived document. Edit it first.")

    def publish(self, doc: Document) -> None:
        print("  [Archived] ERROR: Cannot publish an archived document. Edit it first.")

    def archive(self, doc: Document) -> None:
        print("  [Archived] ERROR: Document is already archived.")


# ---------------------------------------------------------------------------
# Context
# ---------------------------------------------------------------------------
class Document:
    """
    The context object. It delegates state-specific behavior to the current
    state object and allows states to trigger transitions.
    """

    def __init__(self, title: str) -> None:
        self.title = title
        self._state: DocumentState = DraftState()

    @property
    def state_name(self) -> str:
        return self._state.name()

    def set_state(self, state: DocumentState) -> None:
        """Transition to a new state (called by state objects)."""
        old = self._state.name()
        self._state = state
        print(f"  >> Document '{self.title}': {old} -> {self._state.name()}")

    def edit(self) -> None:
        print(f"\nDocument '{self.title}' [{self.state_name}]: edit()")
        self._state.edit(self)

    def submit(self) -> None:
        print(f"\nDocument '{self.title}' [{self.state_name}]: submit()")
        self._state.submit(self)

    def publish(self) -> None:
        print(f"\nDocument '{self.title}' [{self.state_name}]: publish()")
        self._state.publish(self)

    def archive(self) -> None:
        print(f"\nDocument '{self.title}' [{self.state_name}]: archive()")
        self._state.archive(self)


# ---------------------------------------------------------------------------
# Runnable example
# ---------------------------------------------------------------------------
def main() -> None:
    doc = Document("Design Patterns Guide")

    # Normal workflow: Draft -> Review -> Published -> Archived
    doc.edit()       # OK: editing in draft
    doc.publish()    # ERROR: can't publish from draft
    doc.submit()     # OK: moves to Review
    doc.publish()    # OK: moves to Published
    doc.edit()       # ERROR: can't edit published doc
    doc.archive()    # OK: moves to Archived

    # Recover from archived: un-archive by editing
    doc.edit()       # OK: moves back to Draft
    doc.submit()     # OK: moves to Review
    doc.edit()       # OK: rejected, back to Draft (for re-editing)
    doc.submit()     # OK: back to Review
    doc.publish()    # OK: Published again


if __name__ == "__main__":
    main()

File: behavioral-design-pattern/state-pattern/state.ts
/**
 * State Pattern
 * =================
 * Category: Behavioral Design Pattern
 *
 * Intent:
 *   Allow an object to alter its behavior when its internal state changes.
 *   The object will appear to change its class.
 *
 * When to use:
 *   - When an object's behavior depends on its state and it must change
 *     behavior at runtime depending on that state.
 *   - When you want to eliminate large conditional statements that select
 *     behavior based on the current state.
 *   - When transitions between states need to be explicit and controlled.
 *
 * Key Participants:
 *   - State (interface): DocumentState -- declares methods that all concrete
 *     states must implement.
 *   - ConcreteState: DraftState, ReviewState, PublishedState, ArchivedState --
 *     each implements behavior appropriate for that state of the context.
 *   - Context: Document -- maintains a reference to a ConcreteState subclass
 *     that defines the current state.
 */

// ---------------------------------------------------------------------------
// State interface
// ---------------------------------------------------------------------------
interface DocumentState {
  readonly name: string;
  edit(doc: WorkflowDocument): void;
  submit(doc: WorkflowDocument): void;
  publish(doc: WorkflowDocument): void;
  archive(doc: WorkflowDocument): void;
}

// ---------------------------------------------------------------------------
// Concrete states
// ---------------------------------------------------------------------------

class DraftState implements DocumentState {
  readonly name = "Draft";

  edit(doc: WorkflowDocument): void {
    console.log("  [Draft] Editing document content.");
  }

  submit(doc: WorkflowDocument): void {
    console.log("  [Draft] Submitting document for review.");
    doc.setState(new ReviewState());
  }

  publish(doc: WorkflowDocument): void {
    console.log(
      "  [Draft] ERROR: Cannot publish directly from Draft. Submit for review first.",
    );
  }

  archive(doc: WorkflowDocument): void {
    console.log("  [Draft] Archiving draft document.");
    doc.setState(new ArchivedState());
  }
}

class ReviewState implements DocumentState {
  readonly name = "Under Review";

  edit(doc: WorkflowDocument): void {
    console.log("  [Review] Sending document back to Draft for edits.");
    doc.setState(new DraftState());
  }

  submit(doc: WorkflowDocument): void {
    console.log("  [Review] ERROR: Document is already under review.");
  }

  publish(doc: WorkflowDocument): void {
    console.log("  [Review] Review approved. Publishing document.");
    doc.setState(new PublishedState());
  }

  archive(doc: WorkflowDocument): void {
    console.log(
      "  [Review] ERROR: Cannot archive while under review. Publish or reject first.",
    );
  }
}

class PublishedState implements DocumentState {
  readonly name = "Published";

  edit(doc: WorkflowDocument): void {
    console.log(
      "  [Published] ERROR: Cannot edit a published document. Archive it first.",
    );
  }

  submit(doc: WorkflowDocument): void {
    console.log("  [Published] ERROR: Document is already published.");
  }

  publish(doc: WorkflowDocument): void {
    console.log("  [Published] ERROR: Document is already published.");
  }

  archive(doc: WorkflowDocument): void {
    console.log("  [Published] Archiving published document.");
    doc.setState(new ArchivedState());
  }
}

class ArchivedState implements DocumentState {
  readonly name = "Archived";

  edit(doc: WorkflowDocument): void {
    console.log("  [Archived] Un-archiving and moving to Draft for editing.");
    doc.setState(new DraftState());
  }

  submit(doc: WorkflowDocument): void {
    console.log(
      "  [Archived] ERROR: Cannot submit an archived document. Edit it first.",
    );
  }

  publish(doc: WorkflowDocument): void {
    console.log(
      "  [Archived] ERROR: Cannot publish an archived document. Edit it first.",
    );
  }

  archive(doc: WorkflowDocument): void {
    console.log("  [Archived] ERROR: Document is already archived.");
  }
}

// ---------------------------------------------------------------------------
// Context
// ---------------------------------------------------------------------------

class WorkflowDocument {
  private state: DocumentState;

  constructor(public readonly title: string) {
    this.state = new DraftState();
  }

  get stateName(): string {
    return this.state.name;
  }

  /** Transition to a new state (called by state objects). */
  setState(newState: DocumentState): void {
    const old = this.state.name;
    this.state = newState;
    console.log(
      `  >> Document '${this.title}': ${old} -> ${this.state.name}`,
    );
  }

  edit(): void {
    console.log(`\nDocument '${this.title}' [${this.stateName}]: edit()`);
    this.state.edit(this);
  }

  submit(): void {
    console.log(`\nDocument '${this.title}' [${this.stateName}]: submit()`);
    this.state.submit(this);
  }

  publish(): void {
    console.log(`\nDocument '${this.title}' [${this.stateName}]: publish()`);
    this.state.publish(this);
  }

  archive(): void {
    console.log(`\nDocument '${this.title}' [${this.stateName}]: archive()`);
    this.state.archive(this);
  }
}

// ---------------------------------------------------------------------------
// Runnable example
// ---------------------------------------------------------------------------

function main(): void {
  const doc = new WorkflowDocument("Design Patterns Guide");

  // Normal workflow: Draft -> Review -> Published -> Archived
  doc.edit(); // OK: editing in draft
  doc.publish(); // ERROR: can't publish from draft
  doc.submit(); // OK: moves to Review
  doc.publish(); // OK: moves to Published
  doc.edit(); // ERROR: can't edit published doc
  doc.archive(); // OK: moves to Archived

  // Recover from archived: un-archive by editing
  doc.edit(); // OK: moves back to Draft
  doc.submit(); // OK: moves to Review
  doc.edit(); // OK: rejected, back to Draft
  doc.submit(); // OK: back to Review
  doc.publish(); // OK: Published again
}

main();
